// @2bad/bitrix/build/main/bitrix.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = __importDefault(require("./client"));
const companies_1 = __importDefault(require("./services/companies"));
const contacts_1 = __importDefault(require("./services/contacts"));
const deals_1 = __importDefault(require("./services/deals"));
const leads_1 = __importDefault(require("./services/leads"));
const statuses_1 = __importDefault(require("./services/statuses"));
const users_1 = __importDefault(require("./services/users"));
exports.default = (restURI, accessToken) => {
    const { call, batch, list } = client_1.default(restURI, accessToken);
    return {
        call,
        batch,
        list,
        companies: companies_1.default({ call, list }),
        contacts: contacts_1.default({ call, list }),
        deals: deals_1.default({ call, list }),
        leads: leads_1.default({ call, list }),
        statuses: statuses_1.default({ call }),
        users: users_1.default({ call })
    };
};
__export(require("./methods"));
