// @2bad/bitrix/build/main/client/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const got_1 = __importDefault(require("got"));
const p_queue_1 = __importDefault(require("p-queue"));
const addAccessToken_1 = __importDefault(require("./hooks/addAccessToken"));
const batch_1 = __importDefault(require("./methods/batch"));
const call_1 = __importDefault(require("./methods/call"));
const list_1 = __importDefault(require("./methods/list"));
const BITRIX_API_RATE_LIMIT = 2;
const BITRIX_API_RATE_INTERVAL = 1000;
exports.default = (restURI, accessToken) => {
    const client = got_1.default.extend({
        baseUrl: restURI,
        headers: {
            'user-agent': `@2bad/bitrix`
        },
        json: true,
        hooks: {
            beforeRequest: [
                addAccessToken_1.default(accessToken)
            ]
        }
    });
    const queue = new p_queue_1.default({
        intervalCap: BITRIX_API_RATE_LIMIT,
        interval: BITRIX_API_RATE_INTERVAL
    });
    const queuedGet = (...args) => queue.add(() => client.get(...args));
    const call = call_1.default({ get: queuedGet });
    const batch = batch_1.default({ get: queuedGet });
    const list = list_1.default({ call, batch });
    return {
        call,
        batch,
        list
    };
};
