// @2bad/bitrix/build/main/client/methods/list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_range_1 = __importDefault(require("lodash.range"));
const MAX_ENTRIES_PER_COMMAND = 50;
exports.fillWithCommands = ({ method, params }, start, toProcess, entriesPerCommand) => {
    const requiresCommands = Math.ceil((toProcess - start) / entriesPerCommand);
    return lodash_range_1.default(0, requiresCommands)
        .map((i) => ({ method, params: Object.assign(Object.assign({}, params), { start: start + (entriesPerCommand * i) }) }));
};
exports.highest = (input) => Object.values(input).reduce((a, b) => a !== undefined && b !== undefined
    ? a > b ? a : b
    : a !== undefined && b === undefined
        ? a
        : a === undefined && b !== undefined
            ? b
            : undefined, undefined);
exports.batchToListPayload = (payload) => {
    const { result: { result, result_total, result_error, result_next }, time } = payload;
    const flattenResult = Object.entries(result).reduce((flatten, [_key, r]) => !r ? flatten : [...flatten, ...r], []);
    return {
        error: Object.values(result_error).join('\n'),
        next: exports.highest(result_next),
        result: flattenResult,
        time,
        total: exports.highest(result_total) || 0
    };
};
exports.default = ({ call, batch }) => {
    const list = async (method, params) => {
        const start = params.start || 0;
        const listAll = async () => {
            const batchCommands = exports.fillWithCommands({ method, params }, start, firstCall.total, MAX_ENTRIES_PER_COMMAND);
            const payload = await batch(batchCommands);
            return exports.batchToListPayload(payload);
        };
        const firstCall = await call(method, Object.assign(Object.assign({}, params), { start }));
        return !firstCall.next ? firstCall : listAll();
    };
    return list;
};
